### PR TITLE
changes for shallow copy

### PR DESCRIPTION
**Pull request by hemalathagajendran**
## Pull request checklist

```release-notes

```

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature Enhancement
- [ ] Test Automation 
- [ ] Code Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Community Operator listing 
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## How risky is this change?
- [ ] Small, isolated change
- [x] Medium, requires regression testing
- [ ] Large, requires functional and regression testing

### CODE REVIEW

**Review by amdabhad — DISMISSED** [review]

**Comment by hemalathagajendran** [comment]
Design doc: https://ibm.ent.box.com/file/1385848485070

Image: quay.io/hemalatha_gajendran/driver_shallowcopy_feature_latest:latest

**Review by badri-pathak — DISMISSED** [review]

File: driver/csiplugin/controllerserver.go [hunk 1]
```diff
@@ -148,6 +152,10 @@ func (cs *ScaleControllerServer) generateVolID(ctx context.Context, scVol *scale
 		}
 	}
 
+	if isPvcFromSnap{
```
Comment (Jainbrt) [inline driver/csiplugin/controllerserver.go:155]: Should this be one of the if else condition in above code block ?

File: driver/csiplugin/controllerserver.go [hunk 2]
```diff
@@ -665,6 +670,18 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
 		return nil, err
 	}
 
+	isShallowCopyVolume := false
+	for _, reqCap := range reqCapabilities {
+		if reqCap.GetAccessMode().GetMode() == csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY {
```
Comment (Jainbrt) [inline driver/csiplugin/controllerserver.go:675]: should this code be replaced with older code ( line 648-650)?

File: driver/csiplugin/controllerserver.go [hunk 3]
```diff
@@ -702,14 +719,50 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
...
 	if isSnapSource {
-		err = cs.validateSnapId(ctx, scaleVol, &snapIdMembers, scaleVol, primaryClusterID)
+		isValidPvcFromSnapshot, err = cs.validateSnapId(ctx, scaleVol, &snapIdMembers, scaleVol, primaryClusterID, isShallowCopyVolume)
 		if err != nil {
 			klog.Errorf("[%s] volume:[%v] - Error in source snapshot validation [%v]", loggerId, volName, err)
 			return nil, err
 		}
+
+		if !isValidPvcFromSnapshot{
+			return nil, status.Error(codes.Internal, "CreateVolume ValidateShallowCopyVolume failed")
```
Comment (Jainbrt) [inline driver/csiplugin/controllerserver.go:731]: it would a good idea, if we also print a error log message here with volume details.?

**Review by Jainbrt — CHANGES_REQUESTED** [review]

**Review by saurabhwani5 — APPROVED** [review]

**Review by Jainbrt — APPROVED** [review]

**Pull request merged**